c2rust-refactor/scripts/upgrade_ptr_to_ref.lua

```diff
@@ -613,14 +855,17 @@ function Visitor:rewrite_call_expr(expr)
 
                 if #exprs == 1 and path_expr:get_kind() == "Path" then
                     local method_name = param_expr:get_method_name()
+                    local path_cfg = self:get_expr_cfg(path_expr)
 
-                    if method_name == "as_ptr" then
+                    -- If we're looking at an array then we likely don't want
+                    -- a reference to the array type but a raw pointer
+                    if method_name == "as_ptr" and path_cfg and not path_cfg:is_array() then
                         param_expr:to_addr_of(path_expr, false)
-                    elseif method_name == "as_mut_ptr" then
+                    elseif method_name == "as_mut_ptr" and path_cfg and not path_cfg:is_array() then
                         param_expr:to_addr_of(path_expr, true)
                     end
 
-                    if param_cfg:is_opt_any() then
+                    if param_cfg:is_opt_any() and not fn.is_foreign then
```

Comment:
> a local function could still take a pointer

Good point

> A foreign function could take a reference

No? I mean, yes, rust technically allows an extern "C" function to have references in the signature... but does this even make sense? If you're planning to use it in C code then C can't uphold the ref's invariants. If you're planning to use it in rust then you should just have a `use` imported rust function that isn't extern

Reply:
I'm much more concerned about the crate-local function taking a pointer, because they do that after reorg. The other case is unusual, so I'm less concerned about it.